### PR TITLE
Complete Luhn algorithm

### PR DESCRIPTION
**Pull request by cuencandres**
## Change Summary

Luhn algorithm was not complete. An important step of this algorithm was missing.

This PR would solve the problem described in the issue linked below: Luhn Algorithm is not working as expected.

## Related issue number

#1166

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/1166-cuencandres.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by samuelcolvin** [comment]
Looks like build is failing due to an error with netlify, don't worry about that. I'm pretty pissed off with netlify anyway, so will probably move away from them anyway fairly soon, thus not that keen to look into debugging it.

@matin originally implemented `PaymentCardNumber` in #790, @matin can you confirm you agree with this?

**Comment by samuelcolvin** [comment]
Looks good to me.

**Comment by matin** [comment]
@cuencandres  my recommendation is for you to use the `luhnmod10` library like this: https://github.com/cuenca-mx/stpmex-python/blob/master/stpmex/types.py#L156

I made a bad choice implementing luhn 10 from scratch vs using a library that already has a solid implementation.

**Comment by samuelcolvin** [comment]
I'd rather not add another requirement for a single use. From a brief look on wikipedia, the luhn algorithm really doesn't look that complicated. Is there really enough complexity to require a separate library?

**Comment by matin** [comment]
Nope. It's not complex. That's why I had implemented myself at first. I just missed ones part. If you prefer not to have the dependency, I'm ok with the PR as it is now.

**Comment by dmontagu** [comment]
+1 for not adding the dependency

**Comment by samuelcolvin** [comment]
Surely for something with as wide-spread use as this algorithm there should be a pretty comprehensive set of tests which pretty much guarantee a correct algorithm?

**Comment by samuelcolvin** [comment]
Update, let's include all of [these](https://github.com/luhnmod10/python/blob/master/tests.py) test cases in this PR.

**Comment by matin** [comment]
Works for me.

**Comment by matin** [comment]
@cuencandres can you update the tests? Please make sure to use `pytest.mark.parameterize` vs a for loop.

**Comment by cuencandres** [comment]
Sure, I'll work on it.

**Comment by cuencandres** [comment]
I added the same tests @samuelcolvin suggested to include +6 suggested by myself.

File: changes/1166-cuencandres.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+completing Luhn algorithm for ``PaymentCardNumber``.
```
Comment (samuelcolvin) [inline changes/1166-cuencandres.rst]:
```suggestion
completing Luhn algorithm for `PaymentCardNumber`
```

File: tests/test_types_payment_card_number.py [hunk 2]
```diff
@@ -9,11 +9,40 @@
 
 VALID_AMEX = '370000000000002'
 VALID_MC = '5100000000000003'
-VALID_VISA = '4000000000000002'
+VALID_VISA = '4050000000000001'
 VALID_OTHER = '2000000000000000008'
 LUHN_INVALID = '4000000000000000'
 LEN_INVALID = '40000000000000006'
 
+luhn_tests = [
```
Comment (samuelcolvin) [inline tests/test_types_payment_card_number.py]: please move this into the decorator, e.g.

```py
@pytest.mark.parametrize('card_number, valid', [
    ...
])
```

**Comment by samuelcolvin** [comment]
I'm afraid you've messed up the merge with master.

**Comment by samuelcolvin** [comment]
Maybe easiest to start a new pr?

**Comment by cuencandres** [comment]
Sorry, let me create another PR.

**Comment by cuencandres** [comment]
I've fixed the merge. 
I think it is done!

File: tests/test_types_payment_card_number.py [hunk 3]
```diff
@@ -30,10 +31,45 @@ def test_validate_digits():
...
+        ('343728738009847', False),
+        ('000000018', True),
+        ('99999999999999999999', True),
+        ('99999999999999999999999999999999999999999999999999999999999999999997', True),
+    ],
+)
+def test_validate_luhn_check_digit(card_number: str, valid: bool):
+    if valid:
+        assert PaymentCardNumber.validate_luhn_check_digit(card_number) == card_number
+        assert valid
```
Comment (samuelcolvin) [inline tests/test_types_payment_card_number.py]:
```suggestion
```

File: tests/test_types_payment_card_number.py [hunk 4]
```diff
@@ -30,10 +31,45 @@ def test_validate_digits():
...
+    else:
+        with pytest.raises(LuhnValidationError):
+            PaymentCardNumber.validate_luhn_check_digit(card_number)
+            assert not valid
```
Comment (samuelcolvin) [inline tests/test_types_payment_card_number.py]:
```suggestion
```

**Comment by samuelcolvin** [comment]
just one more tiny thing.

By the way, somehow you've made your PR so I can't edit it, hence I can't just fix small issues like this.

In future it might be worthwhile allowing project collaborators to edit your PRs.

**Comment by cuencandres** [comment]
@samuelcolvin , tiny changes are done!

**Pull request merged**